### PR TITLE
`ultralytics 8.3.220` Add ExecuTorch export (.pte) with XNNPACK

### PR DESCRIPTION
**Pull request by ambitious-octopus**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds first-class ExecuTorch support to export and run YOLO11 models on mobile and edge devices, with docs, tests, and CLI/Python integration. 🚀

### 📊 Key Changes
- ExecuTorch export added:
  - New `format=executorch` option creates a `<model>_executorch_model/` folder with `.pte` and `metadata.yaml`.
  - Export implemented with XNNPACK partitioner; requires Python 3.10+, Torch ≥ 2.9.0, `executorch==1.0.0`, and FlatBuffers.
- Inference support:
  - AutoBackend can now load `.pte` files and run inference via the ExecuTorch runtime.
  - CLI/Python flows updated to accept ExecuTorch models for `predict`.
- Documentation and examples:
  - New ExecuTorch integration guide with install, export, mobile/embedded integration, and benchmarks.
  - Export tables and tutorial updated to include ExecuTorch; added to Integrations index and site nav.
- Testing and compatibility:
  - New tests for ExecuTorch export (skipped on Windows; gated by Python/Torch versions).
  - Version bump to `8.3.220`.
- Misc docs polish:
  - Minor copy tweaks (e.g., Seeed Studio reCamera description).

### 🎯 Purpose & Impact
- Seamless edge deployment: One-command export to a portable `.pte` format optimized for on-device inference on iOS, Android, and embedded Linux. 📱🛠️
- Faster CPU performance: Default XNNPACK backend provides strong mobile CPU inference without specialized hardware.
- Better developer experience: Unified CLI/Python workflow for export and predict, structured output with metadata, and clear docs and examples.
- Reliability and CI coverage: New tests ensure export correctness across tasks, improving stability.
- Broader ecosystem support: Enables YOLO11 to fit natively into PyTorch’s edge stack, complementing existing formats like ONNX, TFLite, and TensorRT.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ambitious-octopus, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response — an Ultralytics engineer will review and assist you shortly. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional notes for this PR:
- Please confirm the new `executorch` export target is consistently listed in `export_formats()` and correctly wired in the exporter flow end-to-end.
- Double-check naming consistency (format string, suffix, and user-facing messages) and that the new method is covered by unit tests, including edge cases like dynamic shapes or invalid inputs.
- Ensure dependency checks (`executorch>=0.7.0`) are properly guarded and error messages are clear if the requirement isn’t met.
- If there are known limitations (e.g., fixed input shape), kindly document them in code comments and in the user docs update.
- Consider adding a brief usage snippet to the export docs to align with existing exporter UX, matching the style of other formats.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! ⚡📱

**Comment by onuralpszr** [comment]
The reason we added setuptools because until 0.7.0 there is a usage of "pkg_resources" and it is about to be removed in setuptools and there is a deprecation warning in code as well. But good news is in recent PR it is already fixed. : https://github.com/pytorch/executorch/pull/13513 (it should be available in next release) 

When new package release we can remove setuptools as well. 

cc @ambitious-octopus

**Comment by onuralpszr** [comment]
I also found one left over pkg_resource import in samsung backend : https://github.com/pytorch/executorch/pull/14654
Hopefully it will get merged as well.

**Review by ambitious-octopus — COMMENTED** [review]
@onuralpszr, nice! I'm strarting working on quantization. 
Let's try to have  `XNNPACK Backend` functional for this PR!

**Comment by onuralpszr** [comment]
```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")

executorch_path = model.export(format="executorch")
model = YOLO(executorch_path)

model("https://ultralytics.com/images/bus.jpg",task='detect',imgsz=640, conf=0.25, save=True)
```

@ambitious-octopus , I made some changes on better export and also autobackend part. I also had so make some changes on suffix part to avoid ".pt" vs ".pte" cases. Plus I had to add "yaml" file for metadata for proper "class names" in inference

**Review by ambitious-octopus — COMMENTED** [review]
## YOLO11n + XNNPACK on Raspberry Pi 5

Initial performance tests show significant improvement using ExecuTorch with XNNPACK optimization:

**Results on COCO128:**
- **Baseline PyTorch**: 354.5ms inference
- **ExecuTorch + XNNPACK**: 173.6ms inference  
- **Speedup**: ~2x faster

**Setup Instructions:**

1. Compile ExecuTorch from source (warning: takes 40+ minutes on Pi):

```bash
   # Follow build instructions:
   # https://docs.pytorch.org/executorch/stable/using-executorch-building-from-source.html
```
2. Export model:

```bash
yolo export model=yolo11n.pt format=executorch
```

3. Run validation:
```bash
yolo val model=yolo11n_executorch_model data=coco128.yaml
```

Tested on Raspberry Pi 5. XNNPACK provides ~2x speedup for edge deployment.
cc: @onuralpszr @lakshanthad

**Comment by lakshanthad** [comment]
Great new export addition guys!  @ambitious-octopus @onuralpszr 

Here is performance testing on Raspberry Pi 5 16GB:

Results on COCO128 with YOLO11n:

- Baseline PyTorch: 373.1ms inference
- ExecuTorch + XNNPACK: 184.2ms inference
- Speedup: ~2x faster

Note: It takes about 20 minutes to compile ExecuTorch from source on a Raspberry Pi 5 16GB.

I have also investigated about `aarch64 Linux wheels` for Executorch on [Executorch GitHub](https://github.com/pytorch/executorch) and have come across the following PR which has added the wheel building support on a nightly release cadence but not uploading to PyPi. 
https://github.com/pytorch/executorch/pull/13946

So I have gone ahead and tried the latest nighly release from below with Python3.11 support which comes with Raspberry Pi OS Bookworm (Debian 12):
https://github.com/pytorch/executorch/actions/runs/18128543623

Installation of the above wheel resulted in a missing module:
```sh
ModuleNotFoundError: No module named 'torchao'
```

But I was able to resolve it by `pip install torchao` (torchao has PyPi aarch64 wheel) and then able to successfully export a model with `ExecuTorch`.

So I think we have 2 options here:
1. Host a compiled Executorch aarch64 Linux wheel on our [GitHub assets](https://github.com/ultralytics/assets).
2. Discuss with ExecuTorch guys and maybe Raspberry Pi guys as well to have the official aarch64 Linux wheel support on PyPi.

**Comment by onuralpszr** [comment]
@lakshanthad personally, I wanna see "pypi" push first instead of assets, secondly we need to think about cpp examples because better inference speed and most of the example in their docs also revolving around cpp as well.

**Comment by lakshanthad** [comment]
I understand @onuralpszr. However PyPi push has been discussed in multiple issues and seems the wheels never made it there.

- https://github.com/pytorch/executorch/issues/10651
- https://github.com/pytorch/executorch/issues/7331
- https://github.com/pytorch/executorch/issues/13044

Yeah definitely we can experiment around CPP as well!

**Comment by mergennachin** [comment]
Hi @lakshanthad @onuralpszr @ambitious-octopus

what wheels are missing?

coincidently, we're releasing a new release (1.0) in a few weeks, and we will have a few new wheels, specifically for linux aarch64 in pypi. as of today, it is not published to pypi just yet. 

you can test our release candidate via this command

```
pip install executorch==1.0.0 torch==2.9.0 torchvision==0.24.0 torchao==0.13.0  torchaudio==2.9.0 --extra-index-url https://download.pytorch.org/whl/test/cpu
```

File: ultralytics/nn/autobackend.py [hunk 1]
```diff
@@ -95,6 +95,8 @@ class AutoBackend(nn.Module):
             | NCNN                  | *_ncnn_model/     |
             | IMX                   | *_imx_model/      |
             | RKNN                  | *_rknn_model/     |
+            | Triton Inference      | triton://model    |
+            | Executorch            | *.pte             |
```
Comment (mergennachin) [inline ultralytics/nn/autobackend.py]: nit pick: our official name is ExecuTorch :)

Comment (ambitious-octopus) [reply ultralytics/nn/autobackend.py]: Thanks @mergennachin!

**Comment by mergennachin** [comment]
Carrying this convo from this comment: https://github.com/pytorch/executorch/issues/14644#issuecomment-3349499011

>In our PR we were able to run macos and linux in current version of executorch, also we trying to expand to support multiple backends other than XNNPack, such as mps,coreml,npu etc. Any suggestions also would be very welcome.

My recommendation for next accelerators to try next are would be CoreML, Qnn and Vulkan, and ethosU (for embedded systems).

File: ultralytics/nn/autobackend.py (hunk 1 above)
Comment (onuralpszr) [reply ultralytics/nn/autobackend.py]: All of them has been fixed. Thank you for catching.

**Comment by onuralpszr** [comment]
> Carrying this convo from this comment: [pytorch/executorch#14644 (comment)](https://github.com/pytorch/executorch/issues/14644#issuecomment-3349499011)
> 
> > In our PR we were able to run macos and linux in current version of executorch, also we trying to expand to support multiple backends other than XNNPack, such as mps,coreml,npu etc. Any suggestions also would be very welcome.
> 
> My recommendation for next accelerators to try next are would be CoreML, Qnn and Vulkan, and ethosU (for embedded systems).

Is wheels going to have other backends enabled (for inference) or do we need to re-compile for each of them ?

**Comment by larryliu0820** [comment]
> I understand @onuralpszr. However PyPi push has been discussed in multiple issues and seems the wheels never made it there.
> 
> * [Request for support of ExecuTorch pip package on linux aarch64 pytorch/executorch#10651](https://github.com/pytorch/executorch/issues/10651)
> * [Publish Linux ARM64 Wheels to PyPI pytorch/executorch#7331](https://github.com/pytorch/executorch/issues/7331)
> * [Publish pypi wheel for linux aarch64 pytorch/executorch#13044](https://github.com/pytorch/executorch/issues/13044)
> 
> Yeah definitely we can experiment around CPP as well!

Hi! Have you tried the following:

```
pip install executorch==1.1.0.dev20251001 --extra-index-url https://download.pytorch.org/whl/nightly
```

We will make sure the wheels are pushed to PyPi in our 1.0 release

**Comment by lakshanthad** [comment]
@larryliu0820 I have tried it. A very small issue is that it complains about a missing `torchao` module when trying to export a model. However a simple `pip install torchao` resolves it!

```sh
ModuleNotFoundError: No module named 'torchao'
```

Looking forward to the official PyPi release! Really appreciate the efforts here!

FYI: Experimented on Raspberry Pi 5 16GB (Bookworm OS)

**Comment by mergennachin** [comment]
> . A very small issue is that it complains about a missing torchao module when trying to export a model. 

@lakshanthad yeah, that's expected in our current release candidate. it will be fixed for the official pypi package.

**Comment by mergennachin** [comment]
> Is wheels going to have other backends enabled (for inference) or do we need to re-compile for each of them ?

@onuralpszr

the wheels will have multiple backends enabled (based based on the host machine), so no need to install different packages each time. 

for example Qnn won't be available on MacOS wheel. but we try to bundle majority of the backend by default in pypi

just, this is specifically for model preparation (converting pytorch model to .pte file)

File: ultralytics/nn/autobackend.py [hunk 2]
```diff
@@ -868,19 +894,25 @@ def _model_type(p: str = "path/to/model.pt") -> list[bool]:
         """
         from ultralytics.engine.exporter import export_formats
 
-        sf = export_formats()["Suffix"]  # export suffixes
-        if not is_url(p) and not isinstance(p, str):
-            check_suffix(p, sf)  # checks
-        name = Path(p).name
-        types = [s in name for s in sf]
```
Comment (Y-T-G) [inline ultralytics/nn/autobackend.py:884]: I think we just need to change only this line to `types = [name.endswith(s) for s in name]` to fix the conflict. Since it uses `in` here both `pt` and `pte` matches `.pte`. But with `endswith()`, this shouldn't be an issue.

File: ultralytics/engine/exporter.py [hunk 3]
```diff
@@ -148,6 +150,7 @@ def export_formats():
         ["NCNN", "ncnn", "_ncnn_model", True, True, ["batch", "half"]],
         ["IMX", "imx", "_imx_model", True, True, ["int8", "fraction", "nms"]],
         ["RKNN", "rknn", "_rknn_model", False, False, ["batch", "name"]],
+        ["ExecuTorch", "executorch", "_executorch_model", False, False, ["batch", "name"]],
```
Comment (Y-T-G) [inline ultralytics/engine/exporter.py]: Does `half` or `nms` not work with ExecuTorch?

Comment (onuralpszr) [reply ultralytics/engine/exporter.py]: half is there but unable to do inference (bugs exist) 
NMS is not there , I saw couple different backend also not support yet. (I am researching and checking how to improve)

File: ultralytics/nn/autobackend.py (hunk 2 above)
Comment (onuralpszr) [reply ultralytics/nn/autobackend.py:884]: I already did it in later commits 

```python
    @staticmethod
    def _model_type(p: str = "path/to/model.pt") -> list[bool]:
        """
        Take a path to a model file and return the model type.

        Args:
            p (str): Path to the model file.

        Returns:
            (list[bool]): List of booleans indicating the model type.

        Examples:
            >>> model = AutoBackend(model="path/to/model.onnx")
            >>> model_type = model._model_type()  # returns "onnx"
        """
        from ultralytics.engine.exporter import export_formats

        sf = export_formats()["Suffix"]
        p_path = Path(p)
        if not is_url(p) and not p_path.is_dir():
            check_suffix(p, sf)

        p_str = str(p)
        sf_standard = sf[:-1]  # Standard suffixes from exporter
        pte_suffix = sf[-1]  # '.pte' suffix

        types = [p_str.endswith(s) for s in sf_standard]
        types[5] |= p_str.endswith(".mlmodel")  # CoreML special case
        types[8] &= not types[9]  # TFLite vs. EdgeTPU
        pte = p_str.endswith(pte_suffix)  # ExecuTorch check

        triton = False
        if not (any(types) or pte):
            from urllib.parse import urlsplit

            url = urlsplit(p_str)
            triton = bool(url.netloc) and bool(url.path) and url.scheme in {"http", "grpc"}

        return types + [triton, pte]
```

**Comment by lakshanthad** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `executorch-export` at 2025-10-07 16:32:15 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18319497444): `ci.yml` run 12818
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18319497692): `docker.yml` run 1232

File: ultralytics/engine/exporter.py (hunk 3 above)
Comment (Y-T-G) [reply ultralytics/engine/exporter.py]: @onuralpszr For NMS, you can try wrapping `NMSModel()` before export. Also what about `dynamic`?

File: ultralytics/nn/autobackend.py (hunk 2 above)
Comment (Y-T-G) [reply ultralytics/nn/autobackend.py:884]: @onuralpszr I think it can just be a 1 line change:
`types = [name.endwith(s) for s in sf]`

With this change, the rest of the lines should not be required.

```python
    @staticmethod
    def _model_type(p: str = "path/to/model.pt") -> list[bool]:
        """
        Take a path to a model file and return the model type.

        Args:
            p (str): Path to the model file.

        Returns:
            (list[bool]): List of booleans indicating the model type.

        Examples:
            >>> model = AutoBackend(model="path/to/model.onnx")
            >>> model_type = model._model_type()  # returns "onnx"
        """
        from ultralytics.engine.exporter import export_formats

        sf = export_formats()["Suffix"]  # export suffixes
        if not is_url(p) and not isinstance(p, str):
            check_suffix(p, sf)  # checks
        name = Path(p).name
        types = [name.endswith(s) for s in sf]
        types[5] |= name.endswith(".mlmodel")  # retain support for older Apple CoreML *.mlmodel formats
        types[8] &= not types[9]  # tflite &= not edgetpu
        if any(types):
            triton = False
        else:
            from urllib.parse import urlsplit

            url = urlsplit(p)
            triton = bool(url.netloc) and bool(url.path) and url.scheme in {"http", "grpc"}

        return types + [triton]
 ```

File: ultralytics/engine/exporter.py (hunk 3 above)
Comment (onuralpszr) [reply ultralytics/engine/exporter.py]: @Y-T-G after CI problem, I will test in both inference and expoert sure good idea.

**Comment by lakshanthad** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `executorch-export` at 2025-10-07 16:51:47 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18319974218): `ci.yml` run 12823
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18319974399): `docker.yml` run 1234

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `executorch-export` at 2025-10-08 10:06:41 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18341140239): `ci.yml` run 12843
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18341140412): `docker.yml` run 1236

**Comment by mergennachin** [comment]
@onuralpszr @ambitious-octopus 

we're releasing our pypi packages, hopefully later today

once that is done, do you plan to land this soon? 

if so, we'd love to mention about this integration next week during the pytorch conference and list ultralytics under "ecosystem integration", if that's okay?

**Comment by onuralpszr** [comment]
> @onuralpszr @ambitious-octopus
> 
> we're releasing our pypi packages, hopefully later today
> 
> once that is done, do you plan to land this soon?
> 
> if so, we'd love to mention about this integration next week during the pytorch conference and list ultralytics under "ecosystem integration", if that's okay?

Yes !! If you guys release today I will do test and export and make sure it is relased and we would like to have  "ecosystem integration"  part as well cc @ambitious-octopus

**Comment by mergennachin** [comment]
great, i'll comment here as soon as we publish pypi package today

**Comment by mergennachin** [comment]
Just released 1.0.0 to https://pypi.org/project/executorch/

**Comment by onuralpszr** [comment]
> Just released 1.0.0 to https://pypi.org/project/executorch/

Great, small question still no windows right ? Based on docs ? also "coremltools" version cause a lot of trouble, because how is it versioned did you guys notice or did something about this ?

**Comment by mergennachin** [comment]
@onuralpszr - actually we have windows pypi package as of 1.0.0. just need to update the pypi page too in a bit

**Comment by onuralpszr** [comment]
> @onuralpszr - actually we have windows pypi package as of 1.0.0. just need to update the pypi page too in a bit

okay I also notice some warning related to torchao and torch 2.9.0 support issue, should I skip or force to install 2.8 ?

**Comment by mergennachin** [comment]
> also "coremltools" version cause a lot of trouble, because how is it versioned did you guys notice or did something about this

we haven't noticed. currently, we point to beta version of coreml tools - https://pypi.org/project/coremltools/9.0b1/ for now
https://github.com/pytorch/executorch/blob/release/1.0/pyproject.toml#L76

will update our dependency as soon as they release a production version.

**Comment by mergennachin** [comment]
We currently pin to latest version of torch (2.9) and ao (0.14)

older versions should work most of the time unless we're depending on newer features from ao and torch

**Comment by onuralpszr** [comment]
> We currently pin to latest version of torch (2.9) and ao (0.14)
> 
> older versions should work most of the time unless we're depending on newer features from ao and torch

Great, also for runtime side do we still have same backends enabled ? or more ? 

coreml,mps,xnnpack and core Aten  are what I know

**Comment by mergennachin** [comment]
@onuralpszr We have a few more https://docs.pytorch.org/executorch/main/backends-overview.html

In terms of stability, these are in ready state:

XNNPACK, Core ML, Vulkan, Qualcomm HTP, Ethos-U

**Comment by onuralpszr** [comment]
> @onuralpszr We have a few more https://docs.pytorch.org/executorch/main/backends-overview.html
> 
> In terms of stability, these are in ready state:
> 
> XNNPACK, Core ML, Vulkan, Qualcomm HTP, Ethos-U

Great let me add and test those on export and inference what I can

File: ultralytics/engine/predictor.py [hunk 4]
```diff
@@ -30,6 +30,7 @@
                               yolo11n_ncnn_model         # NCNN
                               yolo11n_imx_model          # Sony IMX
                               yolo11n_rknn_model         # Rockchip RKNN
+                              yolo11n.pte                # PyTorch Executorch
```
Comment (glenn-jocher) [inline ultralytics/engine/predictor.py:33]: Wait is the exported model a `*.pte` file or a `*_executorch_model/` directory?

File: ultralytics/nn/autobackend.py [hunk 5]
```diff
@@ -877,19 +903,25 @@ def _model_type(p: str = "path/to/model.pt") -> list[bool]:
...
+        triton = False
+        if not (any(types) or pte):
             from urllib.parse import urlsplit
 
-            url = urlsplit(p)
+            url = urlsplit(p_str)
             triton = bool(url.netloc) and bool(url.path) and url.scheme in {"http", "grpc"}
 
-        return types + [triton]
+        return types + [triton, pte]
```
Comment (glenn-jocher) [inline ultralytics/nn/autobackend.py]: No, `pte` should be part of `types`, [triton] is concatenated after because it's not a model, it's a URL.

File: ultralytics/nn/autobackend.py (hunk 2 above)
Comment (Y-T-G) [reply ultralytics/nn/autobackend.py:884]: This change doesn't seem to be required now. `model_type()` works for `executorch` without any changes to the original method. Just need to move `pte` above `triton`
https://github.com/ultralytics/ultralytics/blob/a059f5805559a3948a6f28d51f7fa391076175f8/ultralytics/nn/autobackend.py#L182-L183

File: ultralytics/engine/predictor.py (hunk 4 above)
Comment (onuralpszr) [reply ultralytics/engine/predictor.py:33]: It’s a single file, but it includes multiple backends depending on how you export it (XNNPACK, Vulkan, CoreML, MPS, Arm Ethos, Qualcomm, OpenVINO). However, since the file extension is always the same, we wanted to prevent overwriting issues for us and users.

To address this, we can follow the same approach as OpenVINO exports, place each backend in its own folder and include a .yaml file for class names to be used during inference.

File: ultralytics/nn/autobackend.py (hunk 5 above)
Comment (onuralpszr) [reply ultralytics/nn/autobackend.py]: @Y-T-G gave me feedback about ordering so I revert it, that's basically same as it is before 

@Y-T-G thank you !!

**Comment by lakshanthad** [comment]
I have now added benchmarks as well running on Raspberry Pi 5!
We can add more benchmarks later, running on more devices!

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `executorch-export` at 2025-10-21 18:59:45 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18694597058): `ci.yml` run 13456
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18694597071): `docker.yml` run 1317

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `executorch-export` at 2025-10-21 19:32:56 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18695429840): `ci.yml` run 13459
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18695429944): `docker.yml` run 1318

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `executorch-export` at 2025-10-21 19:59:27 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18696019675): `ci.yml` run 13461
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18696019523): `docker.yml` run 1319

**Review by onuralpszr — APPROVED** [review]

**Review by lakshanthad — APPROVED** [review]

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `executorch-export` at 2025-10-21 20:44:17 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/18697115829): `ci.yml` run 13463
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/18697116021): `docker.yml` run 1320

**Comment by onuralpszr** [comment]
@ambitious-octopus @lakshanthad must torch 2.9 because of TorchAO 0.14 needs 2.9 (https://github.com/pytorch/ao/issues/2919) 


image

**Comment by onuralpszr** [comment]
@glenn-jocher  had to make some extra adjustment because of torch 2.9.0 but it is ready for initial merge, tests are should passed, slow CI last minute changes because of docker side was using torch 2.8 make me re-check dep and I make sure executorch using correct packages and make sure match with supported with TorchAO as well. 

cc @ambitious-octopus

**Review by onuralpszr — APPROVED** [review]

**Comment by glenn-jocher** [comment]
Thanks @onuralpszr, looking through this now

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Thanks for the ExecuTorch integration! A few of the new tests gate on Torch versions using string comparisons, which will fail once the minor version reaches double digits (e.g., 2.10). Please switch these skip conditions to use the boolean helpers from `torch_utils` (like `TORCH_2_9`) so the tests continue to run on newer Torch releases.

💬 Posted 3 inline comments

File: tests/test_cli.py [hunk 6]
```diff
@@ -129,3 +129,13 @@ def test_train_gpu(task: str, model: str, data: str) -> None:
 def test_solutions(solution: str) -> None:
     """Test yolo solutions command-line modes."""
     run(f"yolo solutions {solution} verbose=False")
+
+
+@pytest.mark.skipif(
+    not checks.IS_PYTHON_MINIMUM_3_10 or checks.TORCH_VERSION < "2.9.0",
```
Comment (glenn-jocher) [inline tests/test_cli.py]: @onuralpszr this is correct:

```
'2.10.0' < '2.9.0'
Out[2]: True
```

File: ultralytics/engine/exporter.py [hunk 7]
```diff
@@ -1122,6 +1142,39 @@ def export_tflite(self, prefix=colorstr("TensorFlow Lite:")):
             f = saved_model / f"{self.file.stem}_float32.tflite"
         return str(f)
 
+    @try_export
+    def export_executorch(self, prefix=colorstr("ExecuTorch:")):
+        """Exports a model to ExecuTorch (.pte) format into a dedicated directory and saves the required metadata,
+        following Ultralytics conventions.
+        """
+        LOGGER.info(f"\n{prefix} starting export with ExecuTorch...")
+        assert TORCH_2_9, f"ExecuTorch export requires torch>=2.9.0 but torch=={TORCH_VERSION} is installed"
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/exporter.py:1151]: ❗ **CRITICAL**: This assert forces `torch>=2.9`, but PyTorch 2.9 has not been released, so every export attempt (and the new tests) will fail for all users. Please drop the guard to the actual minimum supported version for ExecuTorch (e.g., the version that introduces `torch.export` support) so the exporter is usable.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

ExecuTorch export currently hard-blocks on `torch>=2.9`, which makes the new feature unusable; please relax the version check to the real minimum requirement.

💬 Posted 1 inline comment

File: tests/test_cli.py (hunk 6 above)
Comment (onuralpszr) [reply tests/test_cli.py]: @glenn-jocher  that's probably parsed as 2.1 vs 2.9  also currently torch 2.10 not released. As a suggestion we can fix in different PR as well ?

File: ultralytics/engine/exporter.py (hunk 7 above)
Comment (onuralpszr) [reply ultralytics/engine/exporter.py:1151]: @glenn-jocher ah I see we forced via assert in here

**Review by glenn-jocher — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
🎉 Merged! Huge thanks to @ambitious-octopus for leading [PR #22244](https://github.com/ultralytics/ultralytics/pull/22244), with stellar contributions from @larryliu0820, @lakshanthad, @Y-T-G, @onuralpszr, @mergennachin, and @glenn-jocher. This brings first-class ExecuTorch support to YOLO11—unlocking seamless mobile and edge deployment with a unified CLI/Python experience, solid docs, and reliable tests.

“Innovation is the ability to see change as an opportunity — not a threat.” — Steve Jobs  
Your work turns opportunity into reality: export once, run anywhere, and accelerate CPU inference on-device. 🚀

Quickstart:
- CLI: `yolo export model=yolo11n.pt format=executorch && yolo predict model=yolo11n_executorch_model/model.pte`
- Python:
```python
from ultralytics import YOLO

model = YOLO("yolo11n.pt")
model.export(format="executorch")
YOLO("yolo11n_executorch_model/model.pte")(source="test.jpg")
```

Appreciate the craftsmanship, polish, and care—this meaningfully grows the Ultralytics ecosystem and empowers developers from Ultralytics HUB to embedded devices. Thank you all! 🙌
